rust/libnewsboat/src/config.rs

```diff
@@ -0,0 +1,352 @@
+
+        // in data
+        migrate_file(&newsbeuter_data_dir, &self.m_data_dir, "cache.db").is_ok();
+        migrate_file(&newsbeuter_data_dir, &self.m_data_dir, "queue").is_ok();
+        migrate_file(&newsbeuter_data_dir, &self.m_data_dir, "history.search").is_ok();
+        migrate_file(&newsbeuter_data_dir, &self.m_data_dir, "history.cmdline").is_ok();
+
+        return true;
+    }
+
+    fn migrate_data_from_newsbeuter_simple(&self) -> bool {
+        let newsbeuter_dir = self.m_env_home
+            .join(NEWSBEUTER_CONFIG_SUBDIR);
+
+        if !newsbeuter_dir.is_dir() {
+            return false;
+        }
+
+        let newsboat_dir = self.m_env_home
+            .join(NEWSBOAT_CONFIG_SUBDIR);
+
+        if newsboat_dir.exists() {
+            log!(Level::Debug, "{:?} already exists, aborting migration.", newsboat_dir);
+            return false;
+        }
+
+        if !self.m_silent {
+            eprintln!("Migrating configs and data from Newsbeuter's dotdir...");
+        }
+
+        match mkdir(&newsboat_dir, 0o700) {
+            Ok(_) => (),
+            Err(ref e) if e.kind() == io::ErrorKind::AlreadyExists => (),
+            Err(err) => {
+                if !self.m_silent {
+                    eprintln!("Aborting migration because mkdir on {:?} failed: {}", newsboat_dir, err);
+                }
+                return false;
+            },
+        };
+
+        migrate_file(&newsbeuter_dir, &newsboat_dir, "urls").ok();
+        migrate_file(&newsbeuter_dir, &newsboat_dir, "cache.db").ok();
+        migrate_file(&newsbeuter_dir, &newsboat_dir, "config").ok();
+        migrate_file(&newsbeuter_dir, &newsboat_dir, "queue").ok();
+        migrate_file(&newsbeuter_dir, &newsboat_dir, "history.search").ok();
+        migrate_file(&newsbeuter_dir, &newsboat_dir, "history.cmdline").ok();
+
+        return true;
+    }
+
+    fn migrate_data_from_newsbeuter(&mut self) {
+        let mut migrated = self.migrate_data_from_newsbeuter_xdg();
+
+        if migrated {
+            // Re-running to pick up XDG dirs
+            self.m_url_file = "urls".into();
+            self.m_cache_file = "cache.db".into();
+            self.m_config_file = "config".into();
+            self.m_queue_file = "queue".into();
+            self.find_dirs();
+        } else {
+            migrated = self.migrate_data_from_newsbeuter_simple();
+        }
+
+        if migrated {
+            eprintln!("\nPlease check the results and press Enter to continue.");
+            // only wait for enter, ignore actual input
+            let mut buf = [0; 16];
+            io::stdin().read(&mut buf).ok();
+        }
+    }
+
+    fn create_dirs(&self) -> bool {
+        return try_mkdir(&self.m_config_dir) && try_mkdir(&self.m_data_dir);
+    }
+
+    fn find_dirs(&mut self) {
+        self.m_config_dir = self.m_env_home
+            .join(NEWSBOAT_CONFIG_SUBDIR);
+
+        self.m_data_dir = self.m_config_dir.clone();
```

Comment:
Is cloning necessary here? I noticed there are lots of clone, why not borrow?

Reply:
In this specific place we can't use a borrow because we need the ConfigParser to own the data in m_data_dir, so that the ConfigParser:
1) doesn't outlive it's own data
2) doesn't borrow from itself, which is sadly not supported

Most of the other copies are in getters, I didn't think too much about them since they are currently unused, you are correct that those should be `&str` though.

Reply:
But in this case, why is `self.m_data_dir` same as `self.m_config_dir`? Should not them be equal?

Reply:
@pickfire These two variables are equal when storing everything in dotdir (~/.newsboat), and they aren't equal when storing everything in XDG dirs. `find_dirs()` starts off with an assumption that the dotdir will be used, so it initializes both variables to the same value.

The cloning doesn't bother me because the purpose of this PR is to make a direct port from C++ to Rust. We can polish and improve the Rust version once we fully transitioned and removed the C++ version.

Reply:
Ah, I am interested in the oxiding newsboat as well. Maybe will check out on stuff now and then.

Reply:
Sure! There are a few unfinished tasks in #334 that you can try right now. Otherwise, check on [rust-port label](https://github.com/newsboat/newsboat/issues?q=is%3Aissue+is%3Aopen+label%3Arust-port) from time to time.

Reply:
If you want, can try submitting a `Call for Participation` in this-week-in-rust.org at https://users.rust-lang.org/t/twir-call-for-participation/4821